lib/async/http/faraday/adapter.rb

```diff
@@ -168,22 +173,29 @@ def call(env)
 				
 				def perform_request(env)
 					with_client(env) do |endpoint, client|
+						if headers = env.request_headers
+							headers = ::Protocol::HTTP::Headers[headers]
+							
+							# Use content-length to inform body length if given, but remove the header since it will be
+							# set for us later anyway, and not doing so could result in a duplicate content-length headers
+							# if capitalization differs
+							content_length = headers.delete("content-length")&.to_i
+						end
+						
 						if body = env.body
 							# We need to ensure the body is wrapped in a Readable object so that it can be read in chunks:
 							# Faraday's body only responds to `#read`.
 							if body.is_a?(::Protocol::HTTP::Body::Readable)
 								# Good to go
 							elsif body.respond_to?(:read)
-								body = BodyReadWrapper.new(body)
+								body = BodyReadWrapper.new(body, content_length)
 							else
-								body = ::Protocol::HTTP::Body::Buffered.wrap(body)
+								body = ::Protocol::HTTP::Body::Buffered.wrap(body).then do |buffered|
+									::Protocol::HTTP::Body::Buffered.new(buffered.chunks, content_length)
```

Comment:
This double initialization is ugly but I had no choice since `Protocol::HTTP::Body::Buffered.wrap` doesn't take a `length` argument.

Reply:
I understand how you are thinking about the problem.

A buffered body can compute its size accurately and efficiently from the chunks. That's different from a readable body where we can only call `#read` to stream the data.

Therefore, I wouldn't worry about this code path - just leave it as the original implementation.